# Thống kê mô tả trên tập dữ liệu về phân loại chất lượng rượu đỏ

## 1. Giới thiệu chung

Bộ dữ liệu **Red Wine Quality** được lấy từ nghiên cứu của **Cortez et al. (2009)**, thuộc Đại học Minho (Bồ Đào Nha).  
Dữ liệu mô tả các **đặc tính hóa học** của rượu vang đỏ vùng **Vinho Verde**, và chất lượng cảm quan của chúng (được đánh giá bởi chuyên gia nếm rượu).

## 2. Mục tiêu 


- Thực hiện **thống kê mô tả (Exploratory Data Analysis)** để hiểu cấu trúc dữ liệu.  


## 3. Thông tin về dữ liệu

- **1599 mẫu** rượu vang đỏ  
- **12 biến (features)**: 11 biến độc lập (đặc tính hóa lý) + 1 biến phụ thuộc (quality)

| Biến | Mô tả | Đơn vị | 
|------|--------|---------|
| fixed acidity | Lượng axit cố định (tartaric acid) | g/dm³ | 
| volatile acidity | Lượng axit bay hơi (acetic acid) | g/dm³ | 
| citric acid | Lượng axit citric | g/dm³ | 
| residual sugar | Đường dư sau khi lên men | g/dm³ | 
| chlorides | Lượng muối (clorua) | g/dm³ | 
| free sulfur dioxide | SO₂ ở dạng tự do | mg/dm³ | 
| total sulfur dioxide | Tổng lượng SO₂ | mg/dm³ | 
| density | Mật độ chất lỏng | g/cm³ | 
| pH | Độ pH | - |
| sulphates | Lượng sunphat (kali sunphat) | g/dm³ | 
| alcohol | Nồng độ cồn | % (v/v) |
| quality | Điểm đánh giá chất lượng (0–10) | - |


## 4. Nạp và xem trước dữ liệu

### Import thư viện

In [30]:
import numpy as np
import pandas as pd
from scipy import stats
from scipy.stats import skew, kurtosis

### Đọc dữ liệu

In [5]:
data = pd.read_csv("winequality-red.csv")
data.head(5)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

### Một số thông tin cơ bản của dữ liệu

In [7]:
print("5 dòng đầu: ")
display(data.head(5))

5 dòng đầu: 


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [8]:
print("5 dòng cuối: ")
display(data.tail(5))

5 dòng cuối: 


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6

In [10]:
print("Kiểu dữ liệu của mỗi biến: ")
display(data.dtypes)

Kiểu dữ liệu của mỗi biến: 


fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

In [12]:
print("Số dòng, số cột của dữ liệu: ")
display(data.shape)

Số dòng, số cột của dữ liệu: 


(1599, 12)

In [14]:
print("Kiểm tra dữ liệu và giá trị thiếu: ")
data.info()

Kiểm tra dữ liệu và giá trị thiếu: 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [15]:
print("Kiểm tra giá trị null: ")
data.isnull().sum()

Kiểm tra giá trị null: 


fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

**Nhận xét:** 
Kết quả cho thấy:
- Dữ liệu gồm 12 cột (11 đặc tính + 1 biến mục tiêu `quality`).
- Không có giá trị thiếu (NaN).
- Tất cả biến đều là dạng số (numeric).

## 5. Thống kê mô tả

In [18]:
data.describe().T

count       mean        std      min      25%  \
fixed acidity         1599.0   8.319637   1.741096  4.60000   7.1000   
volatile acidity      1599.0   0.527821   0.179060  0.12000   0.3900   
citric acid           1599.0   0.270976   0.194801  0.00000   0.0900   
residual sugar        1599.0   2.538806   1.409928  0.90000   1.9000   
chlorides             1599.0   0.087467   0.047065  0.01200   0.0700   
free sulfur dioxide   1599.0  15.874922  10.460157  1.00000   7.0000   
total sulfur dioxide  1599.0  46.467792  32.895324  6.00000  22.0000   
density               1599.0   0.996747   0.001887  0.99007   0.9956   
pH                    1599.0   3.311113   0.154386  2.74000   3.2100   
sulphates             1599.0   0.658149   0.169507  0.33000   0.5500   
alcohol               1599.0  10.422983   1.065668  8.40000   9.5000   
quality               1599.0   5.636023   0.807569  3.00000   5.0000   

                           50%        75%        max  
fixed acidity          7.90000   9.200000   15.90000  
volatile acidity       0.52000   0.640000    1.58000  
citric acid            0.26000   0.420000    1.00000  
residual sugar         2.20000   2.600000   15.50000  
chlorides              0.07900   0.090000    0.61100  
free sulfur dioxide   14.00000  21.000000   72.00000  
total sulfur dioxide  38.00000  62.000000  289.00000  
density                0.99675   0.997835    1.00369  
pH                     3.31000   3.400000    4.01000  
sulphates              0.62000   0.730000    2.00000  
alcohol               10.20000  11.100000   14.90000  
quality                6.00000   6.000000    8.00000

**Nhận xét chi tiết theo từng nhóm biến:**
- Nhóm acid
| Biến | Trung bình | Độ lệch chuẩn | Nhận xét |
|------|--------------------|----------------|------------------------------------|
| fixed acidity | 8.32 | 1.74 | Phần lớn rượu có độ acid cố định trung bình, một số mẫu có giá trị cao (~15). |
| volatile acidity | 0.53 | 0.18 | Dao động khá lớn, cho thấy ảnh hưởng đến mùi vị rất khác nhau giữa các mẫu. |
| citric acid | 0.27 | 0.19 | Nhiều mẫu có citric acid thấp, nhưng vài mẫu có giá trị cao tới 1.0. |

- Nhóm chất bảo quản và thành phần:
| Biến | Trung bình | Nhận xét |
|------|--------------------|---------------------------------------------------------|
| residual sugar | 2.54 | Hầu hết rượu khô (ít đường), chỉ vài mẫu có đường cao. |
| chlorides | 0.09 | Hàm lượng muối nhỏ, nhưng nếu vượt 0.2 có thể ảnh hưởng xấu đến vị. |
| free sulfur dioxide | 15.9 | Giá trị dao động rất mạnh (1–72), do kỹ thuật bảo quản khác nhau. |
| total sulfur dioxide | 46.5 | Có mẫu đạt 289, thể hiện mức độ sử dụng SO₂ cao ở vài rượu. |

- Nhóm đặc trưng vật lý:
| Biến | Trung bình | Nhận xét |
|------|--------------------|--------------------------------------------------------------------|
| density | 0.9967| Mật độ tăng khi đường hoặc acid cao → cần xem tương quan với các biến khác. |
| pH |3.31 |Rượu vang đỏ thường chua nhẹ, pH quanh 3.3 là hợp lý. |
| sulphates | 0.66 |Có tác dụng như chất ổn định → có thể ảnh hưởng tích cực đến chất lượng. |
| alcohol | 10.42 |Dao động từ 8.4 đến 14.9, yếu tố quan trọng ảnh hưởng đến cảm nhận. |

- Biến mục tiêu:
| Biến | Trung bình | Ghi chú |
|------|--------------------|--------------------------------------------------------------------|
| quality | 5.64 | Điểm số thường tập trung quanh 5–6, hiếm mẫu cực cao hoặc cực thấp. |

In [29]:
skewness = data.apply(lambda x: skew(x))
kurt = data.apply(lambda x: kurtosis(x))

summary_shape = pd.DataFrame({'Skewness': skewness, 'Kurtosis': kurt})
summary_shape

Skewness   Kurtosis
fixed acidity         0.981829   1.124856
volatile acidity      0.670962   1.217963
citric acid           0.318039  -0.790283
residual sugar        4.536395  28.524438
chlorides             5.675017  41.581708
free sulfur dioxide   1.249394   2.013490
total sulfur dioxide  1.514109   3.794172
density               0.071221   0.927411
pH                    0.193502   0.800671
sulphates             2.426393  11.679884
alcohol               0.860021   0.195654
quality               0.217597   0.292031

**Nhận xét**

| Biến | Nhận xét |
|------|--------------------|
| fixed acidity | Lệch phải nhẹ, phân phối hơi nhọn. |
| volatile acidity | Lệch phải nhẹ, khá gần chuẩn. |
| citric acid | Lệch phải rõ rệt — nhiều mẫu có acid citric thấp. |
| residual sugar | Rất lệch phải, có vài mẫu chứa đường rất cao (outliers). |
| chlorides | Cực lệch phải → nhiều giá trị thấp, ít mẫu muối cao. |
| free sulfur dioxide | Lệch phải nhẹ, có một số ngoại lệ cao. |
| total sulfur dioxide | Lệch phải nhẹ, hơi nhọn. |
| density | Gần chuẩn, ít lệch. |
| pH | Gần chuẩn, phân bố cân đối. |
| sulphates | Lệch phải, có vài mẫu rất cao. |
| alcohol | Lệch phải nhẹ, ổn định. |
| quality | Gần như chuẩn, vì điểm chất lượng rời rạc từ 3–8. |